packages/3dom/src/facade/three-js/correlated-scene-graph.ts

```diff
@@ -0,0 +1,211 @@
+import {Group, Material, Mesh, Object3D, Texture} from 'three';
+import {GLTF as ThreeGLTF, GLTFReference} from 'three/examples/jsm/loaders/GLTFLoader.js';
+
+import {GLTF, GLTFElement} from '../../gltf-2.0.js';
+
+export type ThreeSceneObject = Object3D|Material|Texture;
+type ThreeSceneObjectCallback = (a: ThreeSceneObject, b: ThreeSceneObject) =>
+    void;
+
+export interface GLTFElementHandle {
+  key: Exclude<keyof GLTF, 'scene'>;
+  element: GLTFElement;
+}
+
+export type ThreeObjectSet = Set<ThreeSceneObject>;
+
+export type GLTFElementToThreeObjectMap = Map<GLTFElement, ThreeObjectSet>;
+export type ThreeObjectToGLTFElementHandleMap =
+    Map<Object3D|Material|Texture, GLTFReference>;
+
+const $threeGLTF = Symbol('threeGLTF');
+const $gltf = Symbol('gltf');
+const $gltfElementMap = Symbol('gltfElementMap');
+const $threeObjectMap = Symbol('threeObjectMap');
+const $parallelTraverseThreeScene = Symbol('parallelTraverseThreeScene');
+
+const $correlateOriginalThreeGLTF = Symbol('correlateOriginalThreeGLTF');
+const $correlateCloneThreeGLTF = Symbol('correlateCloneThreeGLTF');
+
+/**
+ * The Three.js GLTFLoader provides us with an in-memory representation
+ * of a glTF in terms of Three.js constructs. It also provides us with a copy
+ * of the deserialized glTF without any Three.js decoration, and a mapping of
+ * glTF elements to their corresponding Three.js constructs.
+ *
+ * A CorrelatedSceneGraph exposes a synchronously available mapping of glTF
+ * element references to their corresponding Three.js constructs.
+ */
+export class CorrelatedSceneGraph {
+  /**
+   * Produce a CorrelatedSceneGraph from a naturally generated Three.js GLTF.
+   * Such GLTFs are produced by Three.js' GLTFLoader, and contain cached
+   * details that expedite the correlation step.
+   *
+   * If a CorrelatedSceneGraph is provided as the second argument, re-correlates
```

Comment:
Is the idea here that if you have a cloned GLTF, you must provide the original CorrelatedSceneGraph because the cached details are no longer useful?

Reply:
Yes, the glTF=>Three.js mappings are by object reference, so they are not applicable to a cloned Three.js graph.